File: py/nth_prime.py
# Willans'_formula - nth_prime.py
# Copyright (c) 2025 epsill0n epsill0n
# Non-commercial use only. Redistribution or modification prohibited without permission.
# See LICENSE.txt for full terms.

import sys
import time
from WF import Willans
nthPrime_seq = Willans.seq.nthPrime
nthPrime_conc = Willans.conc.nthPrime
import mpmath

if __name__=='__main__':
    
    if (len(sys.argv)!=3):
        print(f'Usage: [{sys.argv[0]}] [n] [1/0 for concurrency]')
        exit(-1)
    
    print(f'Using mpmath precision of {mpmath.mp.dps} decimals.')
    
    n = int(sys.argv[1])
    isConcurrent = int(sys.argv[2])

    if(isConcurrent == 0):
        print('Not using concurrency.')
        nthPrime = nthPrime_seq
    else:
        print('Using concurrency.')
        nthPrime = nthPrime_conc

    start = time.time()
    prime = nthPrime(n)
    end = time.time()

    print(prime)

    print(f'Time: {((end - start) * 1e3):.10f}ms')

File: cpp/nth_prime.cpp
// Willans'_formula - nth_prime.cpp
// Copyright (c) 2025 epsill0n epsill0n
// Non-commercial use only. Redistribution or modification prohibited without permission.
// See LICENSE.txt for full terms.

#include <iostream>
#   define FAST_IO
#include <iomanip>

#include <cstdlib>

#   define GCC_O3_OPTIM
#include "WF.hpp"

#include <chrono>


__attribute__((constructor)) void init() {
#ifdef FAST_IO
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(0);
#endif
    WF::init_mpfr();
}


int main(int argc, char** argv) {
    if (argc!=4) {
        std::cout << "Usage: [" << argv[0] << "] [n] [1/0 for concurrency] [1/0 for accumulation]\n" << std::flush;
        return -1;
    }

    std::cout << "Using MPFR precision of " << MPFR_PRECISION << " bits.\n";

    int n = std::atoi(argv[1]);
    bool isConcurrent = bool(std::atoi(argv[2]));
    bool isCumulative = bool(std::atoi(argv[3]));

    mpz_class prime;
    std::chrono::system_clock::time_point begin, end;
    std::chrono::duration<double, std::milli> elapsed;

    if (isConcurrent)
        std::cout << "Using concurrency.\n" << std::flush;
    else std::cout << "Not using concurrency.\n" << std::flush;

    if (isCumulative)
        std::cout << "Using accumulation.\n" << std::flush;
    else std::cout << "Not using accumulation.\n" << std::flush;

    WF::init_cumulative(isConcurrent, isCumulative);
    
    if (isConcurrent) {

        begin = std::chrono::high_resolution_clock::now();
        prime = WF::conc::nthPrime(n);
        end = std::chrono::high_resolution_clock::now();
        
    } else {

        begin = std::chrono::high_resolution_clock::now();
        prime = WF::seq::nthPrime(n);
        end = std::chrono::high_resolution_clock::now();

    }

    elapsed = end - begin;
    std::cout << prime << "\n";
    std::cout << "Time: " << std::fixed << std::setprecision(10) <<  (elapsed.count()) << "ms\n";
    

    std::cout << std::flush;
}